File: hw1/turtlejoy/src/turtlejoy.cpp
#include <geometry_msgs/Twist.h>
#include <ros/ros.h>
#include <ros/console.h>
#include <sensor_msgs/Joy.h>

class TurtleJoy {
  public:
    TurtleJoy();
  private:
    void joyCallback(const sensor_msgs::Joy::ConstPtr& joy);

    ros::NodeHandle nh;

    ros::Publisher vel_pub;
    ros::Subscriber joy_sub;

    int l_axis_pos = 5;
    int l_axis_neg = 2;
    int a_axis = 0;

    double l_scale = 1.0;
    double a_scale = 1.0;
};

TurtleJoy::TurtleJoy() {
  ROS_INFO("Starting turtlejoy node with name %s", ros::this_node::getName().c_str());

  vel_pub = nh.advertise<geometry_msgs::Twist>("/vrep/drone/cmd_vel", 1);
  joy_sub = nh.subscribe<sensor_msgs::Joy>("joy", 10, &TurtleJoy::joyCallback, this);
}

void TurtleJoy::joyCallback(const sensor_msgs::Joy::ConstPtr& joy) {
  geometry_msgs::Twist twist;
  twist.linear.x = l_scale * (joy->axes[l_axis_neg] - joy->axes[l_axis_pos]) / 2.0;
  // twist.angular.z = a_scale * joy->axes[a_axis];

  ROS_DEBUG("[%s]: vel_x: %f omega_z: %f", ros::this_node::getName().c_str(), twist.linear.x, twist.angular.z);

  vel_pub.publish(twist);
}

int main(int argc, char** argv) {
  ros::init(argc, argv, "turtlejoy");
  if( ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME, ros::console::levels::Debug) ) {
    ros::console::notifyLoggerLevelsChanged();
  }
  TurtleJoy turtlejoy;

  ros::spin();
}

File: hw1/turtlejoy/scripts/turtlejoy.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Joy

class TurtleJoy:    
    l_axis_pos = 5
    l_axis_neg = 2
    a_axis = 0

    l_scale = 1.0
    a_scale = 1.0

    def joyCallback(self, data):
        twist = Twist()
        twist.linear.x = self.l_scale * (data.axes[self.l_axis_neg] - data.axes[self.l_axis_pos]) / 2.0
        twist.angular.z = self.a_scale * data.axes[self.a_axis] 

        rospy.logdebug("{}: vel_x: {: .2f} omega_z: {: .2f}".format(rospy.get_name(), twist.linear.x, twist.angular.z))
        self.vel_pub.publish(twist)

    def __init__(self):
        rospy.loginfo("Starting turtlejoy node with name {}".format(rospy.get_name()))

        self.joy_sub = rospy.Subscriber("/joy", Joy, self.joyCallback)
        self.vel_pub = rospy.Publisher("/vrep/twistCommand", Twist, queue_size=1)

if __name__ == '__main__':
    rospy.init_node("turtlejoy", log_level=rospy.DEBUG)
    tj = TurtleJoy()
    rospy.spin()
